### PR TITLE
feat(rule-engine): add update resource support

### PR DESCRIPTION
**Pull request by wwhai**

### CODE REVIEW

**Comment by turtleDeng** [comment]
Please add test cases.

File: test/emqx_rule_engine_SUITE.erl [hunk 1]
```diff
@@ -443,12 +443,49 @@ t_crud_resources_api(_Config) ->
...
+    %
+    {ok, #{code := 0}} = emqx_rule_engine_api:update_resource(#{id => ResId},
+                                                              [{<<"config">>, [{<<"a">>, 2}]},
+                                                               {<<"description">>, <<"2">>}]),
+    {ok, #{code := 0, data := Resources3}} = emqx_rule_engine_api:show_resource(#{id => ResId},[]),
+    ?assertEqual(ResId, maps:get(id, Resources3)),
+    ?assertEqual(#{<<"a">> => 2}, maps:get(config, Resources3)),
+    ?assertEqual(<<"2">>, maps:get(description, Resources3)),
+    {ok, #{code := 0, data := Resources3}} = emqx_rule_engine_api:show_resource(#{id => ResId},[]),
+    ?assertEqual(ResId, maps:get(id, Resources3)),
```
Comment (turtleDeng) [inline test/emqx_rule_engine_SUITE.erl]: Code duplication

File: src/emqx_rule_engine.erl [hunk 2]
```diff
@@ -289,23 +347,25 @@ delete_resource(ResId) ->
...
-     catch _:Error:ST ->
-        logger:critical(
-            "Can not re-stablish resource ~p: ~0p. The resource is disconnected."
-            "Fix the issue and establish it manually.\n"
-            "Stacktrace: ~0p",
-            [ResId, Error, ST])
-     end || Res = #resource{id = ResId}
-            <- emqx_rule_registry:get_resources()],
-    ok.
+    lists:foreach(fun(Resource) ->
```
Comment (hjianbo) [inline src/emqx_rule_engine.erl]:
```suggestion
    lists:foreach(fun refresh_resource/1,
```

File: src/emqx_rule_engine.erl [hunk 3]
```diff
@@ -289,23 +347,25 @@ delete_resource(ResId) ->
...
+        refresh_resource(Resource)
+    end, emqx_rule_registry:get_resources()).
 
 refresh_resource(Type) when is_atom(Type) ->
-    [refresh_resource(Resource)
-     || Resource <- emqx_rule_registry:get_resources_by_type(Type)];
+    lists:foreach(fun(Resource) ->
```
Comment (hjianbo) [inline src/emqx_rule_engine.erl]: lists:foreach(fun refresh_resource/1,

File: src/emqx_rule_engine_api.erl [hunk 4]
```diff
@@ -325,6 +333,33 @@ show_resource(#{id := Id}, _Params) ->
...
+            return(ok);
+        {error, not_found} ->
+            ?LOG(error, "resource not found: ~0p", [Id]),
+            return({error, 400, <<"resource not found:", Id/binary>>});
+        {error, {init_resource_failure, _}} ->
+            ?LOG(error, "init resource failure: ~0p", [Id]),
+            return({error, 500, <<"init resource failure:", Id/binary>>});
+        {error, {dependency_exists, RuleId}} ->
+            ?LOG(error, "dependency exists: ~0p", [RuleId]),
+            return({error, 500, <<"resource dependency by rule:", RuleId/binary>>});
```
Comment (hjianbo) [inline src/emqx_rule_engine_api.erl]: Please fix this error message

File: src/emqx_rule_engine_api.erl [hunk 5]
```diff
@@ -325,6 +333,33 @@ show_resource(#{id := Id}, _Params) ->
...
+        {error, Reason} ->
+            ?LOG(error, "update resource failed: ~0p", [Reason]),
+            return({error, 500, <<"update resource failed,error info have been written to logfile!">>})
```
Comment (hjianbo) [inline src/emqx_rule_engine_api.erl]: Please fix this error message

File: src/emqx_rule_engine_cli.erl [hunk 6]
```diff
@@ -166,6 +172,18 @@ resources(["create" | Params]) ->
...
 
+resources(["update" | Params]) ->
+    with_opts(fun({Opts, _}) ->
+        Id = proplists:get_value(id, Opts),
+        Maps = make_updated_resource(Opts),
+        case emqx_rule_engine:update_resource(Id, Maps) of
+            ok ->
+                emqx_ctl:print("Resource update successfully~n");
+            {error, Reason} ->
+                emqx_ctl:print("update resource failed, reason: ~p!~n", [Reason])
```
Comment (hjianbo) [inline src/emqx_rule_engine_cli.erl]:
```suggestion
                emqx_ctl:print("Resource update failed: ~0p~n", [Reason])
```

File: test/emqx_rule_engine_SUITE.erl [hunk 7]
```diff
@@ -443,12 +443,47 @@ t_crud_resources_api(_Config) ->
...
     ?assert(length(Resources) > 0),
-
     {ok, #{code := 0, data := Resources2}} = emqx_rule_engine_api:show_resource(#{id => ResId},[]),
     ?assertEqual(ResId, maps:get(id, Resources2)),
-
+    %
+    {ok, #{code := 0}} = emqx_rule_engine_api:update_resource(#{id => ResId},
+                                                              [{<<"config">>, [{<<"a">>, 2}]},
+                                                               {<<"description">>, <<"2">>}]),
+    {ok, #{code := 0, data := Resources3}} = emqx_rule_engine_api:show_resource(#{id => ResId},[]),
```
Comment (hjianbo) [inline test/emqx_rule_engine_SUITE.erl]:
```suggestion
    {ok, #{code := 0, data := Resources3}} = emqx_rule_engine_api:show_resource(#{id => ResId}, []),
```

**Review by HJianBo — CHANGES_REQUESTED** [review]

File: src/emqx_rule_engine_api.erl (hunk 4 above)
Comment (hjianbo) [reply src/emqx_rule_engine_api.erl]: 4.3.0 also needs to be modified.

**Review by HJianBo — APPROVED** [review]

**Pull request merged**